instrumentation/github.com/aws/aws-lambda-go/otellambda/lambda.go

```diff
@@ -0,0 +1,351 @@
+	}
+	return true
+}
+
+type wrappedHandler struct {
+	handler lambda.Handler
+}
+
+// Compile time check our Handler implements lambda.Handler
+var _ lambda.Handler = wrappedHandler{}
+
+// Invoke adds OTel span surrounding customer Handler invocation
+func (h wrappedHandler) Invoke(ctx context.Context, payload []byte) ([]byte, error) {
+
+	ctx, span := tracingBegin(ctx)
+	defer tracingEnd(ctx, span)
+
+	response, err := h.handler.Invoke(ctx, payload)
+	if err != nil {
+		return nil, err
+	}
+
+	return response, nil
+}
+
+// HandlerWrapper Provides a Handler which wraps customer Handler with OTel Tracing
+func HandlerWrapper(handler lambda.Handler) lambda.Handler {
+	return wrappedHandler{handler: handler}
+}
+
+// Logic to start OTel Tracing
+func tracingBegin(ctx context.Context) (context.Context, trace.Span) {
+	// Add trace id to context
+	xrayTraceId := os.Getenv("_X_AMZN_TRACE_ID")
+	mc := mapCarrier{}
+	mc.Set("X-Amzn-Trace-Id", xrayTraceId)
+	propagator := xray.Propagator{}
+	ctx = propagator.Extract(ctx, mc)
+
+	// If tracer provider initialization failed we
+	// will attempt to initialize once per invocation
+	if tp == nil {
+		initTracerProvider()
+	}
+
+	// if tracer provider successfully initializes then
+	// we add tracing, otherwise do customer business
+	// logic with no tracing
+	if tp != nil {
+		// Get a named tracer with package path as its name.
+		tracer := tp.Tracer(tracerName)
+
+		var span trace.Span
+		spanName := os.Getenv("AWS_LAMBDA_FUNCTION_NAME")
+
+		var attributes []attribute.KeyValue
+		lc, ok := lambdacontext.FromContext(ctx)
+		if !ok {
+			errorLogger.Println("failed to load lambda context from context, ensure tracing enabled in Lambda")
+		}
+		if lc != nil {
+			ctxRequestID := lc.AwsRequestID
+			attributes = append(attributes, attribute.KeyValue{Key: semconv.FaaSExecutionKey, Value: attribute.StringValue(ctxRequestID)})
+
+			// Resource attrs added as span attr due to static tp
+			// being created without meaningful context
+			ctxFunctionArn := lc.InvokedFunctionArn
+			attributes = append(attributes, attribute.KeyValue{Key: semconv.FaaSIDKey, Value: attribute.StringValue(ctxFunctionArn)})
+			arnParts := strings.Split(ctxFunctionArn, ":")
+			if len(arnParts) >= 5 {
+				attributes = append(attributes, attribute.KeyValue{Key: semconv.CloudAccountIDKey, Value: attribute.StringValue(arnParts[4])})
+			}
+		}
+
+		ctx, span = tracer.Start(ctx, spanName, trace.WithSpanKind(trace.SpanKindServer), trace.WithAttributes(attributes...))
+
+		return ctx, span
+	}
+	return ctx, nil
+}
+
+// Logic to wrap up OTel Tracing
+func tracingEnd(ctx context.Context, span trace.Span) {
+	if tp != nil {
+		// span will be valid if tp is not nil
+		span.End()
+
+		// yield processor to attempt to attempt to ensure
+		// all spans have been consumed and are ready to be
+		// flushed - see https://github.com/open-telemetry/opentelemetry-go/issues/2068
+		// to be removed upon resolution of above issue
+		runtime.Gosched()
```

Comment:
This would only yield to another goroutine with no guarantees that it will be the flusher. I think we need to think a bit more what we should do here. Ideally, a graceful shutdown event would be ideal place to force flush but this doesn't exist at the moment.

Can we write a custom batcher for Lambda that flushes in the same goroutine and document the limitation?

Reply:
I talked to the Lambda team and they said the extensions API is supported in the runtime/functions: https://docs.aws.amazon.com/lambda/latest/dg/runtimes-extensions-api.html#runtimes-extensions-api-lifecycle

We should be able to listen to the "shutdown" event and force flush there instead of force flushing after each handler. The shutdown happens after the Lambda VM is recycled and this may add up to 500ms of latency but I think we are fine with this.

The Lambda collector is implementing something to listen to these events: https://github.com/open-telemetry/opentelemetry-lambda/blob/main/collector/extension/client.go You can use that as a reference.

Reply:
@rakyll Refer [design](https://github.com/open-telemetry/opentelemetry-lambda/blob/main/docs/design_proposal.md#21-lambda-freeze)
1. adding force flush at the end of each Lambda invocation but not in `shutdown` is for no data delay.
2. Collector monitor Lambda event does not make sense. Because Collector flush data does not mean SDK has flushed data to Collector.

Reply:
> This would only yield to another goroutine with no guarantees that it will be the flusher. I think we need to think a bit more what we should do here. Ideally, a graceful shutdown event would be ideal place to force flush but this doesn't exist at the moment.
> 
> Can we write a custom batcher for Lambda that flushes in the same goroutine and document the limitation?

The issue should be addressed by https://github.com/open-telemetry/opentelemetry-go/issues/2080. We list some workarounds before fixing ForceFlush():
1. As your suggestion, create a specific batch processor for Lambda only // difficulty? more maintenance; useless after fixing ForceFlush;
2. Using SimpleProcessor instead of BatchProcessor  // bad performance, ie: create 100 spans in one Lambda function.
3. Add a yield, give up CPU to other gorountine, and hopefully Enqueue can clean the data on the way.

If in actual testing the `yield` can solve problem, it maybe the best option.